server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -522,150 +529,229 @@ public IndexMetaDataGenerations indexMetaDataGenerations() {
+        for (Map.Entry<SnapshotId, Map<String, String>> snapshotIdMapEntry : indexMetaLookup.entrySet()) {
+            final Map<String, String> val = snapshotIdMapEntry.getValue();
+            final Map<IndexId, String> forSnapshot = new HashMap<>(val.size());
+            for (Map.Entry<String, String> generationEntry : val.entrySet()) {
+                forSnapshot.put(indexLookup.get(generationEntry.getKey()), generationEntry.getValue());
+            }
+            indexGenerations.put(snapshotIdMapEntry.getKey(), forSnapshot);
+        }
+        return new IndexMetaDataGenerations(indexGenerations, indexMetaIdentifiers);
+    }
+
+    /**
+     * Parses the "snapshots" field and fills maps for the various per snapshot properties. This method must run before
+     * {@link #parseIndices} which will rely on the maps of snapshot properties to have been populated already.
+     *
+     * @param parser           x-content parse
+     * @param snapshots        map of snapshot uuid to {@link SnapshotId}
+     * @param snapshotStates   map of snapshot uuid to {@link SnapshotState}
+     * @param snapshotVersions map of snapshot uuid to {@link Version} that the snapshot was taken in
+     * @param indexMetaLookup  map of {@link SnapshotId} to map of index id (as returned by {@link IndexId#getId}) that defines the index
+     *                         metadata generations for the snapshot
+     */
+    private static void parseSnapshots(XContentParser parser, Map<String, SnapshotId> snapshots, Map<String, SnapshotState> snapshotStates,
+                                       Map<String, Version> snapshotVersions,
+                                       Map<SnapshotId, Map<String, String>> indexMetaLookup) throws IOException {
+        if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+            throw new ElasticsearchParseException("expected array for [" + SNAPSHOTS + "]");
+        }
+        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+            String name = null;
+            String uuid = null;
+            SnapshotState state = null;
+            Map<String, String> metaGenerations = null;
+            Version version = null;
+            while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                String currentFieldName = parser.currentName();
+                parser.nextToken();
+                switch (currentFieldName) {
+                    case NAME:
+                        name = parser.text();
+                        break;
+                    case UUID:
+                        uuid = parser.text();
+                        break;
+                    case STATE:
+                        state = SnapshotState.fromValue((byte) parser.intValue());
+                        break;
+                    case INDEX_METADATA_LOOKUP:
+                        metaGenerations = parser.mapStrings();
```

Comment:
Again saving a map relative to the old implementation by not using `putAll` once on a single map and instead just using the map from the parser